### ISSUE TITLE
Strategy for dynamic list

### ISSUE THREAD

**Issue by matt-way** [alexbrillant/react-native-deck-swiper#153]
I am using firebase, and updating properties of items on swipe that would cause them not be rendered next time. This issue is that this is screwing up the index.

For example, the swiper initialises with 5 cards. When the first card is swiped, a property on that object is changed which propagates back, causing a re-render, however this time because of the update there is only 4 cards (the first is no longer in the list). This causes the swiper to skip the current card, because the index is now 1, but the index should change back to zero.

Any ideas for how to get around this?

Just to recap, the issue is removing the first card in the list on every swipe causes the indexes to stuff up.

**Comment by webraptor** [comment]
You can set the _cardIndex_ on your parent component whenever you want the deck to re-render, as in your case, when you update a property on some card. 

```
const { cardIndex } = this.props;
return (<Swiper
ref={swiper => {
  {...customSwiperProps}
  cardIndex={cardIndex}
/>
```

Passing along the cardIndex to the swiper will allow you to modify it (aka reset or whatever) externally, thus triggering a re-render of the deck of cards. All onSwipe callbacks return the cardIndex, so you can use that to push the updated cardIndex to your state (redux or whatever you're using).

**Comment by alexbrillant** [comment]
should we add this to the readme ? should we close this issue ?

**Comment by webraptor** [comment]
Yeah it might be useful in the readme since info about what happens when card content changes is not displayed anywhere.

**Comment by vinayr** [comment]
@webraptor is the code supposed to be like this?
```
const { cardIndex } = this.props;
return (
  <Swiper
    ref={swiper => this.swiper = swiper}
    {...customSwiperProps}
    cardIndex={cardIndex}
  />
);
```

**Comment by webraptor** [comment]
@vinayr Yes! nice catch. Must have miss deleted some bits here and there.

**Comment by noman-10p** [comment]
Thanks, what If I change any value in object externally of the current card,does it automatically re-renders ?

**Comment by webraptor** [comment]
Yes. shouldComponentUpdate checks whether cards have changed or not.

**Comment by tehwayne** [comment]
Is it possible to customize shouldComponentUpdate? I update cards on swipe, however would like to prevent re-rendering of the cards until I restart the swiper from the start

**Comment by webraptor** [comment]
Well shouldComponentUpdate does exactly what it's supposed to do. Prevent re-rendering of cards unless they actually change. Perhaps share the code so that we can help you out? The other solution would be to fork the project so that you customise in your own way the method.

**Comment by tehwayne** [comment]
@webraptor fair enough, I'll try a check in the parent component instead

**Comment by webraptor** [comment]
@matt-way / @alexbrillant / @tehwayne Is this issue still present, or can we close this down?

**Comment by pierregoutheraud** [comment]
Passing cardIndex re-renders the cards but it does not seem to fix the index problem.

My swiper initializes with a list [card1, card2, card3]. 
As soon as I swiped card1, I am removing it from the list, causing a re-render of the swiper with the updated list [card2, card3].
The problem is that after the first swipe and the re-render, the swiper displays card3 instead of card2.

I do not see how keeping track and updating cardIndex changes that...

![](https://d16rm1n165bd05.cloudfront.net/items/0x0f041R3K3H3r1u2T0i/Screen%20Recording%202019-02-22%20at%2006.26%20PM.gif?v=aa5e3587?raw=true)

**Comment by acrabb** [comment]
Not sure if this is the same base issue, but when I get to the end of my list of cards, sometimes I add more cards (update the underlying array for the `cards` prop in `onSwipedAll`).
However, the card deck isn't re-rendering with the new list. The last card gets swiped away and then its blank.

**Comment by grazianospinelli** [comment]
> 
> 
> You can set the _cardIndex_ on your parent component whenever you want the deck to re-render, as in your case, when you update a property on some card.
> 
> ```
> const { cardIndex } = this.props;
> return (<Swiper
>     ref={ swiper => this.swiper = swiper }
>     {...customSwiperProps}
>     cardIndex={cardIndex}
>   />
> );```
> 
> Passing along the cardIndex to the swiper will allow you to modify it (aka reset or whatever) externally, thus triggering a re-render of the deck of cards. All onSwipe callbacks return the cardIndex, so you can use that to push the updated cardIndex to your state (redux or whatever you're using).
> ```

Sorry for my noobness! I really do not understand how to use this piece of code. I have built another component called MySwiper with that code inside... and called it from parent component but i have not understood how to pass {...customSwiperProps}.  They are undefined in parent. And should do i use ref prop in parent too?

**Comment by dekameron22** [comment]
> Not sure if this is the same base issue, but when I get to the end of my list of cards, sometimes I add more cards (update the underlying array for the `cards` prop in `onSwipedAll`).
> However, the card deck isn't re-rendering with the new list. The last card gets swiped away and then its blank.

+1

**Comment by gurgensimonyan** [comment]
> Not sure if this is the same base issue, but when I get to the end of my list of cards, sometimes I add more cards (update the underlying array for the `cards` prop in `onSwipedAll`).
> However, the card deck isn't re-rendering with the new list. The last card gets swiped away and then its blank.

I had the same issue. When I tried to debug the source code, I found out that every time the stack is being rendered with cards property from state object, not from props object, which means that if initially you provide 5 cards, later it doesn't matter your card list will contain more or less cards, it still will render the stack with 5 cards. I forked the repo and changed that part, so now card list is updating dynamically.

**Comment by BTCBoss** [comment]
I am passing updated cards to the swiper and it's not updating. Is this issue going to be fixed? If I update the cards..i don't want to update the index...the state of the index should be maintained by the swiper

**Comment by eamonnalphin** [comment]
I can't figure out how to make this work, I can create a deck, swipe it away, update the underlying array, force update the component that has the Swiper in it, but the deck stays gone, the cards don't show up again. Changing the card index doesn't do anything, am I missing something?

**Comment by peterkortvel** [comment]
Hi, would anybody be able to add an example on how to handle this? Do I need to build another parent component or how do I pass the index and props so it updates? Thank you so much!

**Comment by hyochan** [comment]
> Not sure if this is the same base issue, but when I get to the end of my list of cards, sometimes I add more cards (update the underlying array for the `cards` prop in `onSwipedAll`).
> However, the card deck isn't re-rendering with the new list. The last card gets swiped away and then its blank.

For those of you who are still having an issue, I've created [PR](https://github.com/alexbrillant/react-native-deck-swiper/pull/340) for a workaround. Hope this helps.

**Comment by servonlewis** [comment]
> > Not sure if this is the same base issue, but when I get to the end of my list of cards, sometimes I add more cards (update the underlying array for the `cards` prop in `onSwipedAll`).
> 
> > However, the card deck isn't re-rendering with the new list. The last card gets swiped away and then its blank.
> 
> 
> 
> For those of you who are still having an issue, I've created [PR](https://github.com/alexbrillant/react-native-deck-swiper/pull/340) for a workaround. Hope this helps.

Is there a way I can use this now? Maybe show me how to download it pointing to your branch?

Or show me how to use the npm module patch thingy. I can't seem to get this go work, using react functional components.

**Comment by webraptor** [comment]
The PR is made against the repo that is no longer maintained. Current npm releases are made from https://github.com/webraptor/react-native-deck-swiper so please open PRs against that one

**Comment by servonlewis** [comment]
> The PR is made against the repo that is no longer maintained. Current npm releases are made from https://github.com/webraptor/react-native-deck-swiper so please open PRs against that one

It looks like that one has the same npm install name. How would npm know which one to install?  Or was the npm module swapped to the new one already?

**Comment by hyochan** [comment]
> > > Not sure if this is the same base issue, but when I get to the end of my list of cards, sometimes I add more cards (update the underlying array for the `cards` prop in `onSwipedAll`).
> > 
> > 
> > > However, the card deck isn't re-rendering with the new list. The last card gets swiped away and then its blank.
> > 
> > 
> > For those of you who are still having an issue, I've created [PR](https://github.com/alexbrillant/react-native-deck-swiper/pull/340) for a workaround. Hope this helps.
> 
> Is there a way I can use this now? Maybe show me how to download it pointing to your branch?
> 
> Or show me how to use the npm module patch thingy. I can't seem to get this go work, using react functional components.

You can just install this like below in your `package.json`.
```
"react-native-deck-swiper": "git+https://github.com/collectors-dev/react-native-deck-swiper.git#1.0.0",
```

**Comment by hyochan** [comment]
> The PR is made against the repo that is no longer maintained. Current npm releases are made from https://github.com/webraptor/react-native-deck-swiper so please open PRs against that one

Will do thanks.

**Comment by webraptor** [comment]
> > The PR is made against the repo that is no longer maintained. Current npm releases are made from https://github.com/webraptor/react-native-deck-swiper so please open PRs against that one
> 
> It looks like that one has the same npm install name. How would npm know which one to install? Or was the npm module swapped to the new one already?

Yes. npm was switched a few months back. You can double check the homepage / repository links from https://www.npmjs.com/package/react-native-deck-swiper

**Comment by sirajalam049** [comment]
> > The PR is made against the repo that is no longer maintained. Current npm releases are made from https://github.com/webraptor/react-native-deck-swiper so please open PRs against that one
> 
> Will do thanks.

Just wanted to know if is this done? As the feature still not working.

**Comment by sirajalam049** [comment]
> > > > Not sure if this is the same base issue, but when I get to the end of my list of cards, sometimes I add more cards (update the underlying array for the `cards` prop in `onSwipedAll`).
> > > 
> > > 
> > > > However, the card deck isn't re-rendering with the new list. The last card gets swiped away and then its blank.
> > > 
> > > 
> > > For those of you who are still having an issue, I've created [PR](https://github.com/alexbrillant/react-native-deck-swiper/pull/340) for a workaround. Hope this helps.
> > 
> > 
> > Is there a way I can use this now? Maybe show me how to download it pointing to your branch?
> > Or show me how to use the npm module patch thingy. I can't seem to get this go work, using react functional components.
> 
> You can just install this like below in your `package.json`.
> 
> ```
> "react-native-deck-swiper": "git+https://github.com/collectors-dev/react-native-deck-swiper.git#1.0.0",
> ```

This is throwing error ViewOverflow not defined. Also no typings.

**Comment by webraptor** [comment]
> > > > > Not sure if this is the same base issue, but when I get to the end of my list of cards, sometimes I add more cards (update the underlying array for the `cards` prop in `onSwipedAll`).
> > > > 
> > > > 
> > > > > However, the card deck isn't re-rendering with the new list. The last card gets swiped away and then its blank.
> > > > 
> > > > 
> > > > For those of you who are still having an issue, I've created [PR](https://github.com/alexbrillant/react-native-deck-swiper/pull/340) for a workaround. Hope this helps.
> > > 
> > > 
> > > Is there a way I can use this now? Maybe show me how to download it pointing to your branch?
> > > Or show me how to use the npm module patch thingy. I can't seem to get this go work, using react functional components.
> > 
> > 
> > You can just install this like below in your `package.json`.
> > ```
> > "react-native-deck-swiper": "git+https://github.com/collectors-dev/react-native-deck-swiper.git#1.0.0",
> > ```
> 
> This is throwing error ViewOverflow not defined. Also no typings.

That's probably a fork of a older version of the package. version 2.0.4 removed react-native-view-overflow from the source and this error no longer occurs.

**Comment by webraptor** [comment]
> > > The PR is made against the repo that is no longer maintained. Current npm releases are made from https://github.com/webraptor/react-native-deck-swiper so please open PRs against that one
> > 
> > 
> > Will do thanks.
> 
> Just wanted to know if is this done? As the feature still not working.

There were no open PRs for this particular issue. However, properly managing the swiper parent's state as described above allows updating the swiper without issues.
